### ISSUE TITLE
TransformerEncoderLayer produce different result with padding

### ISSUE THREAD

**Issue by ronzhou** [pytorch/pytorch#90534]
### 🐛 Describe the bug

```python
import torch

d_model = 10
n_head = 5
batch_size = 5
_record = 50

torch.manual_seed(0)
_encoder_layer = nn.TransformerEncoderLayer(d_model=d_model, nhead=n_head, batch_first=True)
torch.manual_seed(0)
feature_value = torch.randn(batch_size, _record, d_model)

_mask = torch.ones(batch_size, _record)
_mask[:, -1] = 0.0             # mask the last row for each batch
    
_encoder_layer.eval()
# prediction include the last row but mask it
result1 = _encoder_layer(feature_value, src_key_padding_mask=_mask)    
# prediction without mask and exclude the last row
result2 = _encoder_layer(feature_value[:, :-1, :])           

print(result1[0, 0])
print(result2[0, 0])    
"""
tensor([-0.8711, -0.8349,  0.0806,  0.4240,  1.5906,  1.3947, -0.1622, -1.5281,
         0.8241, -0.9176], grad_fn=<SelectBackward0>)
tensor([-0.8746, -0.8340,  0.0819,  0.4208,  1.5915,  1.3968, -0.1670, -1.5278,
         0.8238, -0.9116], grad_fn=<SelectBackward0>)
"""
```
Ideally, this two cases should produce exactly same result, but they are different. If I change the _record value bigger, like 5k, the difference shrink. I also checked if adding src_mask will aviod the difference, but it produce the same result without scr_mask. I'm wondering if this is the desired result.

Thanks for your help!

### Versions

PyTorch version: 1.12.0+cu116
Is debug build: False
CUDA used to build PyTorch: 11.6
ROCM used to build PyTorch: N/A

OS: CentOS Stream 9 (x86_64)
GCC version: (GCC) 11.3.1 20220421 (Red Hat 11.3.1-2)
Clang version: 14.0.5 (Red Hat 14.0.5-1.el9)
CMake version: version 3.20.2
Libc version: glibc-2.10

Python version: 3.7.6 (default, Jan  8 2020, 19:59:22)  [GCC 7.3.0] (64-bit runtime)
Python platform: Linux-5.14.0-142.el9.x86_64-x86_64-with-centos-9
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: 
GPU models and configuration: 
GPU 0: NVIDIA GeForce RTX 3070
GPU 1: NVIDIA GeForce RTX 3070
GPU 2: NVIDIA GeForce RTX 3070
GPU 3: NVIDIA GeForce RTX 3070
GPU 4: NVIDIA GeForce RTX 3070

Nvidia driver version: 515.43.04
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.19.5
[pip3] numpydoc==0.9.2
[pip3] torch==1.12.0+cu116
[pip3] torchaudio==0.12.0+cu116
[pip3] torchvision==0.13.0+cu116
[conda] blas                      1.0                         mkl    conda-forge
[conda] mkl                       2021.2.0                 pypi_0    pypi
[conda] mkl-service               2.3.0            py37he904b0f_0  
[conda] mkl_fft                   1.0.15           py37ha843d7b_0  
[conda] mkl_random                1.1.0            py37hd6b4f25_0  
[conda] numpy                     1.18.1           py37h4f9e942_0  
[conda] numpy-base                1.18.1           py37hde5b4d6_1  
[conda] numpydoc                  0.9.2                      py_0    conda-forge

cc @jbschlosser @bhosmer @cpuhrsch @erichan1

**Comment by ronzhou** [comment]
I think the difference comes from the multiattention part. According to the paper "Attention Is All You Need“, the attention matrix will be divided by sqrt(d_k), which is the sequence len. So adding a padding record will change the value of d_k, thus affect the final result. I think the more reasonable way is to divide the matrix by the sqrt of the length of the unpadding data.

**Comment by Vasud-ha** [comment]
This issue is reproducible.

**Comment by ronzhou** [comment]
I made a mistate about src_key_padding_mask. The mask matrix should take True as padding value.
